core/src/avm2/globals/flash/display/graphics.rs

```diff
@@ -364,6 +365,365 @@ pub fn draw_rect<'gc>(
+            // point.
+
+            let unit_r_br_curve_x = unit_r_point_x + a_b * unit_r_point_y * -1.0;
+            let unit_r_br_curve_y = unit_r_point_y + a_b * unit_r_point_x;
+
+            // We need another Bezier segment, of course...
+
+            let unit_b_point_x = (PI * 0.5).cos();
+            let unit_b_point_y = (PI * 0.5).sin();
+
+            let unit_br_b_curve_x = unit_br_point_x + a_b * unit_br_point_y * -1.0;
+            let unit_br_b_curve_y = unit_br_point_y + a_b * unit_br_point_x;
+
+            // At this point we have 90 degrees of a circle in Bezier points.
+            // The remaining points can be derived by coordinate flips,
+            // folllowed by scaling to get an ellipse, and translation to the
+            // actual ellipse centers. All the scary trig is over with.
+            //
+            // We'll start from the bottom-right corner of the rectangle,
+            // because that's what Flash Player does.
+
+            let line_width = width - ellipse_width;
+            let line_height = height - ellipse_height;
+
+            let br_ellipse_center_x = x + ellipse_width / 2.0 + line_width;
+            let br_ellipse_center_y = y + ellipse_height / 2.0 + line_height;
+
+            let br_point_x = br_ellipse_center_x + ellipse_width / 2.0 * unit_br_point_x;
+            let br_point_y = br_ellipse_center_y + ellipse_height / 2.0 * unit_br_point_y;
+
+            draw.draw_command(DrawCommand::MoveTo {
+                x: Twips::from_pixels(br_point_x),
+                y: Twips::from_pixels(br_point_y),
+            });
+
+            let br_b_curve_x = br_ellipse_center_x + ellipse_width / 2.0 * unit_br_b_curve_x;
+            let br_b_curve_y = br_ellipse_center_y + ellipse_height / 2.0 * unit_br_b_curve_y;
+
+            let right_b_point_x = br_ellipse_center_x + ellipse_width / 2.0 * unit_b_point_x;
+            let right_b_point_y = br_ellipse_center_y + ellipse_height / 2.0 * unit_b_point_y;
+
+            draw.draw_command(DrawCommand::CurveTo {
+                x1: Twips::from_pixels(br_b_curve_x),
+                y1: Twips::from_pixels(br_b_curve_y),
+                x2: Twips::from_pixels(right_b_point_x),
+                y2: Twips::from_pixels(right_b_point_y),
+            });
+
+            // Oh, since we're drawing roundrects, we also need to draw lines
+            // in between each ellipse. This is the bottom line.
+            let tl_ellipse_center_x = x + ellipse_width / 2.0;
+            let tl_ellipse_center_y = y + ellipse_height / 2.0;
+
+            let left_b_point_x = tl_ellipse_center_x + ellipse_width / -2.0 * unit_b_point_x;
+            let left_b_point_y = br_ellipse_center_y + ellipse_height / 2.0 * unit_b_point_y;
+
+            draw.draw_command(DrawCommand::LineTo {
+                x: Twips::from_pixels(left_b_point_x),
+                y: Twips::from_pixels(left_b_point_y),
+            });
+
+            // Bottom-left ellipse
```

Comment:
Could this repetition be replaced with a loop? 4-side loop would be nice, but I imagine 2-side loop would be simpler to write (due to trivial symmetry)

Reply:
I don't think so, because there's *just enough* of a difference in the generated command set to make this impractical.

First, we have to start the command set with a `MoveTo`, unlike all the other commands. Second, we start at the bottom-right corner of the rect, *not* at the start or end of any ellipses. This is primarily motivated by the fact that I plan to eventually implement `readGraphicsData` and use it to provide tests for all the `Graphics` methods - which would require generating the same shape data Flash does.